### DISCUSSION TITLE
详解 Kubernetes 垃圾收集器的实现原理 · /kubernetes-garbage-collector

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#548] [Announcements]
https://draveness.me/kubernetes-garbage-collector

**Comment by xinnjie** [comment]
> 在 Kubernetes 引入垃圾收集器之前，所有的级联删除逻辑都是在客户端完成的，kubectl 会先删除 ReplicaSet 持有的 Pod 再删除 ReplicaSet

在引入垃圾收集器之前，kubectl 要做这么多事吗？。  以我理解就算没有垃圾收集器，也应该交给 ReplicaSet controller 做级联删除的操作。垃圾收集器的引入大大减少built-in controller 和 自定义controoler实现级联删除的负担。

**Comment by draveness** [comment]
> @xinnjie
> > 在 Kubernetes 引入垃圾收集器之前，所有的级联删除逻辑都是在客户端完成的，kubectl 会先删除 ReplicaSet 持有的 Pod 再删除 ReplicaSet
> 
> 在引入垃圾收集器之前，kubectl 要做这么多事吗？。  以我理解就算没有垃圾收集器，也应该交给 ReplicaSet controller 做级联删除的操作。垃圾收集器的引入大大减少built-in controller 和 自定义controoler实现级联删除的负担。

看一下 Kubernetes 垃圾收集器的 Proposal 

> Currently most cascading deletion logic is implemented at client-side. For example, when deleting a replica set, kubectl uses a reaper to delete the created pods and then delete the replica set. We plan to move the cascading deletion to the server to simplify the client-side logic. In this proposal, we present the garbage collector which implements cascading deletion for all API resources in a generic way; we also present the finalizer framework, particularly the "orphan" finalizer, to enable flexible alternation between cascading deletion and orphaning. [^1]

[^1]: Cascading deletion with Garbage Collector <https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/garbage-collection.md#overview>
